File: Implementations/Algorithms/Sort_Algorithms/insertion_sort.cpp
#include <iostream>

using namespace std;

void insertionSort(int arr[], int n){
    int j, swaps = 0, comps = 0;
     for(j = 1; j < n; j++){
        int i = j - 1; 
        int key = arr[j];
        
        while(i >= 0 && arr[i] > key){
            swaps++;
            arr[i + 1] = arr[i];
            i--;
        }
        comps++;
        arr[i + 1] = key;
        
        for(int k = 0; k < n; k++) printf("Arr[%d] = %d\t", k, arr[k]);
        cout << endl;
    }
    printf("\nComparisons Made: %d\n", comps);
    printf("Element moves Made: %d\n", swaps);
}

void printArray(int arr[], int n)
{
    int i;
    for (i = 0; i < n; i++)
        cout << arr[i] << " ";
    cout << endl;
}

int main() {
    // int arr[] = { 4, 3, 2, 10, 12, 1, 5, 6 };
    // int arr[] = { 10, 20, 30, 40, 50, 60 };
    int arr[] = { 60, 50, 40, 30, 20, 10 };
    int n = sizeof(arr) / sizeof(arr[0]);
 
    insertionSort(arr, n);
    printArray(arr, n);
    return 0;
}

File: Javascript_DSALGO/Sort Algorithms/insertionSort.js
function insertionSort(arr) {
  for (var i = 1; i < arr.length; i++) {
    var currentVal = arr[i];
    for (var j = i - 1; j >= 0 && arr[j] > currentVal; j--) {
      arr[j + 1] = arr[j];
    }
    arr[j + 1] = currentVal;
  }
  return arr;
}

function insertionSort(arr, comp) {
  if (!comp) comp = (a, b) => a - b;
  for (var i = 1; i < arr.length; i++) {
    var currentVal = arr[i];
    for (var j = i - 1; j >= 0 && comp(arr[j], currentVal) > 0; j--) {
      arr[j + 1] = arr[j];
    }
    arr[j + 1] = currentVal;
  }
  return arr;
}
